File: Coding4Fun.Obd/Coding4Fun.Obd.InstrumentClusterVB/Classes/InstrumentClusterSkin.vb
Imports System.Windows
Imports System.Windows.Controls

Imports Coding4Fun.Obd.InstrumentCluster.Controls
Imports Coding4Fun.Obd.InstrumentCluster.Controls.Skin1967

Namespace Coding4Fun.Obd.InstrumentCluster.Classes
	Public Class InstrumentClusterSkin
		Inherits UserControl
		Public uxSettings As Button
		Public uxLeftTurn As Blinker
		Public uxRightTurn As Blinker
		Public uxLowFuel As FrameworkElement
		Public uxHazard As FrameworkElement
		Public uxMalfunction As FrameworkElement
		Public uxOpenDoor As FrameworkElement
		Public uxMessage As TextBlock
		Public uxMessageMarquee As Marquee
		Public uxOdometer As TextBlock
		Public uxMPG As TextBlock
		Public uxMPH As Slider
		Public uxRPM As Slider
		Public uxFuel As Slider
		Public uxTemp As Slider
		Public uxOil As Slider
		Public uxBattery As Slider
		Public uxMPHLabel As TextBlock
		Public uxRPMLabel As TextBlock

		Private _IsLeftTurnBlinkerVisible As Boolean
		Private _IsRightTurnBlinkerVisible As Boolean
		Private _IsLowFuelVisible As Boolean
		Private _IsHazardVisible As Boolean
		Private _IsMalfunctionVisible As Boolean
		Private _IsOpenDoorVisible As Boolean
		Private _Message As String
		Private _Mileage As Double
		Private _MPG As Double
		Private _MPH As Double
		Private _RPM As Double
		Private _Fuel As Double
		Private _Temperature As Double
		Private _Oil As Double
		Private _BatteryLevel As Double

		Public Sub New()
			AddHandler Initialized, AddressOf Skin_Initialized
		End Sub

		Private Sub Skin_Initialized(ByVal sender As Object, ByVal e As EventArgs)
			uxSettings = TryCast(FindName("uxSettings"), Button)
			uxLeftTurn = TryCast(FindName("uxLeftTurn"), Blinker)
			uxRightTurn = TryCast(FindName("uxRightTurn"), Blinker)
			uxLowFuel = TryCast(FindName("uxLowFuel"), FrameworkElement)
			uxHazard = TryCast(FindName("uxHazard"), FrameworkElement)
			uxMalfunction = TryCast(FindName("uxMalfunction"), FrameworkElement)
			uxOpenDoor = TryCast(FindName("uxOpenDoor"), FrameworkElement)
			uxMessage = TryCast(FindName("uxMessage"), TextBlock)
			uxMessageMarquee = TryCast(FindName("uxMessageMarquee"), Marquee)
			uxOdometer = TryCast(FindName("uxOdometer"), TextBlock)
			uxMPG = TryCast(FindName("uxMPG"), TextBlock)
			uxMPH = TryCast(FindName("uxMPH"), Slider)
			uxRPM = TryCast(FindName("uxRPM"), Slider)
			uxFuel = TryCast(FindName("uxFuel"), Slider)
			uxTemp = TryCast(FindName("uxTemp"), Slider)
			uxOil = TryCast(FindName("uxOil"), Slider)
			uxBattery = TryCast(FindName("uxBattery"), Slider)
			uxMPHLabel = TryCast(FindName("uxMPHLabel"), TextBlock)
			uxRPMLabel = TryCast(FindName("uxRPMLabel"), TextBlock)
		End Sub

		Public Property IsLeftTurnBlinkerVisible() As Boolean
			Get
				Return _IsLeftTurnBlinkerVisible
			End Get
			Set(ByVal value As Boolean)
				_IsLeftTurnBlinkerVisible = value
				If uxLeftTurn IsNot Nothing Then
					uxLeftTurn.Visibility = If(value, Visibility.Visible, Visibility.Hidden)
				End If
			End Set
		End Property

		Public Property IsRightTurnBlinkerVisible() As Boolean
			Get
				Return _IsRightTurnBlinkerVisible
			End Get
			Set(ByVal value As Boolean)
				_IsRightTurnBlinkerVisible = value
				If uxRightTurn IsNot Nothing Then
					uxRightTurn.Visibility = If(value, Visibility.Visible, Visibility.Hidden)
				End If
			End Set
		End Property

		Public Property IsLowFuelVisible() As Boolean
			Get
				Return _IsLowFuelVisible
			End Get
			Set(ByVal value As Boolean)
				_IsLowFuelVisible = value
				If uxLowFuel IsNot Nothing Then
					uxLowFuel.Visibility = If(value, Visibility.Visible, Visibility.Hidden)
				End If
				If uxFuel IsNot Nothing Then
					uxFuel.IsEnabled = Not value
				End If
			End Set
		End Property

		Public Property IsHazardVisible() As Boolean
			Get
				Return _IsHazardVisible
			End Get
			Set(ByVal value As Boolean)
				_IsHazardVisible = value
				If uxHazard IsNot Nothing Then
					uxHazard.Visibility = If(value, Visibility.Visible, Visibility.Hidden)
				End If
			End Set
		End Property

		Public Property IsMalfunctionVisible() As Boolean
			Get
				Return _IsMalfunctionVisible
			End Get
			Set(ByVal value As Boolean)
				_IsMalfunctionVisible = value
				If uxMalfunction IsNot Nothing Then
					uxMalfunction.Visibility = If(value, Visibility.Visible, Visibility.Hidden)
				End If
			End Set
		End Property

		Public Property IsOpenDoorVisible() As Boolean
			Get
				Return _IsOpenDoorVisible
			End Get
			Set(ByVal value As Boolean)
				_IsOpenDoorVisible = value
				If uxOpenDoor IsNot Nothing Then
					uxOpenDoor.Visibility = If(value, Visibility.Visible, Visibility.Hidden)
				End If
			End Set
		End Property

		Public Property Message() As String
			Get
				Return _Message
			End Get
			Set(ByVal value As String)
				_Message = value
				If uxMessage IsNot Nothing Then
					uxMessage.Text = value
				End If
				If uxMessageMarquee IsNot Nothing Then
					uxMessageMarquee.Text = value
				End If
			End Set
		End Property

		Public Property Mileage() As Double
			Get
				Return _Mileage
			End Get
			Set(ByVal value As Double)
				_Mileage = value
				If uxOdometer IsNot Nothing Then
					uxOdometer.Text = Math.Floor(value).ToString()
				End If
			End Set
		End Property

		Public Property MPG() As Double
			Get
				Return _MPG
			End Get
			Set(ByVal value As Double)
				_MPG = value
				If uxMPG IsNot Nothing Then
					uxMPG.Text = Math.Round(value).ToString()
				End If
			End Set
		End Property

		Public Property MPH() As Double
			Get
				Return _MPH
			End Get
			Set(ByVal value As Double)
				_MPH = value
				If uxMPH IsNot Nothing Then
					uxMPH.Value = value
				End If
				If uxMPHLabel IsNot Nothing Then
					uxMPHLabel.Text = Math.Ceiling(value).ToString()
				End If
			End Set
		End Property

		Public Property RPM() As Double
			Get
				Return _RPM
			End Get
			Set(ByVal value As Double)
				_RPM = value
				If uxRPM IsNot Nothing Then
					uxRPM.Value = value
				End If
				If uxRPMLabel IsNot Nothing Then
					uxRPMLabel.Text = Math.Ceiling(value).ToString()
				End If
			End Set
		End Property

		Public Property Fuel() As Double
			Get
				Return _Fuel
			End Get
			Set(ByVal value As Double)
				_Fuel = value
				If uxFuel IsNot Nothing Then
					uxFuel.Value = value
				End If
				IsLowFuelVisible = value < 10
			End Set
		End Property

		Public Property Temperature() As Double
			Get
				Return _Temperature
			End Get
			Set(ByVal value As Double)
				_Temperature = value
				If uxTemp IsNot Nothing Then
					uxTemp.Value = value
				End If
			End Set
		End Property

		Public Property Oil() As Double
			Get
				Return _Oil
			End Get
			Set(ByVal value As Double)
				_Oil = value
				If uxOil IsNot Nothing Then
					uxOil.Value = value
				End If
			End Set
		End Property

		Public Property BatteryLevel() As Double
			Get
				Return _BatteryLevel
			End Get
			Set(ByVal value As Double)
				_BatteryLevel = value
				If uxBattery IsNot Nothing Then
					uxBattery.Value = value
				End If
			End Set
		End Property
	End Class
End Namespace

File: Coding4Fun.Obd/Coding4Fun.Obd.InstrumentCluster/Classes/InstrumentClusterSkin.cs
using System;
using System.Windows.Controls;
using System.Windows;
using Coding4Fun.Obd.InstrumentCluster.Controls;
using Coding4Fun.Obd.InstrumentCluster.Controls.Skin1967;

namespace Coding4Fun.Obd.InstrumentCluster.Classes
{
    public class InstrumentClusterSkin : UserControl
    {
        public Button uxSettings;
        public Blinker uxLeftTurn;
        public Blinker uxRightTurn;
        public FrameworkElement uxLowFuel;
        public FrameworkElement uxHazard;
        public FrameworkElement uxMalfunction;
        public FrameworkElement uxOpenDoor;
        public TextBlock uxMessage;
        public Marquee uxMessageMarquee;
        public TextBlock uxOdometer;
        public TextBlock uxMPG;
        public Slider uxMPH;
        public Slider uxRPM;
        public Slider uxFuel;
        public Slider uxTemp;
        public Slider uxOil;
        public Slider uxBattery;
        public TextBlock uxMPHLabel;
        public TextBlock uxRPMLabel;

        private bool _IsLeftTurnBlinkerVisible;
        private bool _IsRightTurnBlinkerVisible;
        private bool _IsLowFuelVisible;
        private bool _IsHazardVisible;
        private bool _IsMalfunctionVisible;
        private bool _IsOpenDoorVisible;
        private string _Message;
        private double _Mileage;
        private double _MPG;
        private double _MPH;
        private double _RPM;
        private double _Fuel;
        private double _Temperature;
        private double _Oil;
        private double _BatteryLevel;

        public InstrumentClusterSkin()
        {
            Initialized += new EventHandler(Skin_Initialized);
        }

        void Skin_Initialized(object sender, EventArgs e)
        {
            uxSettings = FindName("uxSettings") as Button;
            uxLeftTurn = FindName("uxLeftTurn") as Blinker;
            uxRightTurn = FindName("uxRightTurn") as Blinker;
            uxLowFuel = FindName("uxLowFuel") as FrameworkElement;
            uxHazard = FindName("uxHazard") as FrameworkElement;
            uxMalfunction = FindName("uxMalfunction") as FrameworkElement;
            uxOpenDoor = FindName("uxOpenDoor") as FrameworkElement;
            uxMessage = FindName("uxMessage") as TextBlock;
            uxMessageMarquee = FindName("uxMessageMarquee") as Marquee;
            uxOdometer = FindName("uxOdometer") as TextBlock;
            uxMPG = FindName("uxMPG") as TextBlock;
            uxMPH = FindName("uxMPH") as Slider;
            uxRPM = FindName("uxRPM") as Slider;
            uxFuel = FindName("uxFuel") as Slider;
            uxTemp = FindName("uxTemp") as Slider;
            uxOil = FindName("uxOil") as Slider;
            uxBattery = FindName("uxBattery") as Slider;
            uxMPHLabel = FindName("uxMPHLabel") as TextBlock;
            uxRPMLabel = FindName("uxRPMLabel") as TextBlock;
        }

        public bool IsLeftTurnBlinkerVisible
        {
            get { return _IsLeftTurnBlinkerVisible; }
            set
            {
                _IsLeftTurnBlinkerVisible = value;
                if (uxLeftTurn != null) uxLeftTurn.Visibility = value ? Visibility.Visible : Visibility.Hidden;
            }
        }

        public bool IsRightTurnBlinkerVisible
        {
            get { return _IsRightTurnBlinkerVisible; }
            set
            {
                _IsRightTurnBlinkerVisible = value;
                if (uxRightTurn != null) uxRightTurn.Visibility = value ? Visibility.Visible : Visibility.Hidden;
            }
        }

        public bool IsLowFuelVisible
        {
            get { return _IsLowFuelVisible; }
            set
            {
                _IsLowFuelVisible = value;
                if (uxLowFuel != null) uxLowFuel.Visibility = value ? Visibility.Visible : Visibility.Hidden;
                if (uxFuel != null) uxFuel.IsEnabled = !value;
            }
        }

        public bool IsHazardVisible
        {
            get { return _IsHazardVisible; }
            set
            {
                _IsHazardVisible = value;
                if (uxHazard != null) uxHazard.Visibility = value ? Visibility.Visible : Visibility.Hidden;
            }
        }

        public bool IsMalfunctionVisible
        {
            get { return _IsMalfunctionVisible; }
            set
            {
                _IsMalfunctionVisible = value;
                if(uxMalfunction != null) uxMalfunction.Visibility = value ? Visibility.Visible : Visibility.Hidden;
            }
        }

        public bool IsOpenDoorVisible
        {
            get { return _IsOpenDoorVisible; }
            set
            {
                _IsOpenDoorVisible = value;
                if (uxOpenDoor != null) uxOpenDoor.Visibility = value ? Visibility.Visible : Visibility.Hidden;
            }
        }

        public string Message
        {
            get { return _Message; }
            set
            {
                _Message = value;
                if (uxMessage != null) uxMessage.Text = value;
                if (uxMessageMarquee != null) uxMessageMarquee.Text = value;
            }
        }

        public double Mileage
        {
            get { return _Mileage; }
            set
            {
                _Mileage = value;
                if (uxOdometer != null) uxOdometer.Text = Math.Floor(value).ToString();
            }
        }

        public double MPG
        {
            get { return _MPG; }
            set
            {
                _MPG = value;
                if (uxMPG != null) uxMPG.Text = Math.Round(value).ToString();
            }
        }

        public double MPH { get { return _MPH; } 
            set {
                _MPH = value;
                if(uxMPH != null) uxMPH.Value = value;
                if (uxMPHLabel != null) uxMPHLabel.Text = Math.Ceiling(value).ToString();
            }
        }

        public double RPM { get { return _RPM; } 
            set {
                _RPM = value;
                if (uxRPM != null) uxRPM.Value = value;
                if (uxRPMLabel != null) uxRPMLabel.Text = Math.Ceiling(value).ToString();
            }
        }

        public double Fuel
        {
            get { return _Fuel; }
            set
            {
                _Fuel = value;
                if (uxFuel != null) uxFuel.Value = value;
                IsLowFuelVisible = value < 10;
            }
        }

        public double Temperature
        {
            get { return _Temperature; }
            set
            {
                _Temperature = value;
                if (uxTemp != null) uxTemp.Value = value;
            }
        }

        public double Oil
        {
            get { return _Oil; }
            set
            {
                _Oil = value;
                if (uxOil != null) uxOil.Value = value;
            }
        }

        public double BatteryLevel
        {
            get { return _BatteryLevel; }
            set
            {
                _BatteryLevel = value;
                if (uxBattery != null) uxBattery.Value = value;
            }
        }
    }
}
